src/rdb.c

```diff
@@ -2185,6 +2217,162 @@ robj *rdbLoadObject(int rdbtype, rio *rdb, sds key, int dbid, int *error) {
+            if (expire != 0 && iAmMaster() && ((mstime_t)expire < now) &&  /* note: expire was saved to RDB as unix-time in milliseconds */
+                !(rdbflags & RDBFLAGS_AOF_PREAMBLE)) {
+                /* TODO: consider replication (like in rdbLoadAddKeyToDb) */
+                server.rdb_last_load_hash_fields_expired++;
+                sdsfree(field);
+                sdsfree(value);
+                continue;
+            }
+
+            /* keep the nearest expiration to connect listpack object to db expiry */
+            if ((expire != 0) && (expire < minExpire)) minExpire = expire;
+
+            /* store the values read - either to listpack or dict */
+            if (o->encoding == OBJ_ENCODING_LISTPACK_EX) {
+                /* integrity - check for key duplication (if required) */
+                if (dupSearchDict) {
+                    sds field_dup = sdsnewlen(field, sdslen(field));
+
+                    if (dictAdd(dupSearchDict, field_dup, NULL) != DICT_OK) {
+                        rdbReportCorruptRDB("Hash with dup elements");
+                        dictRelease(dupSearchDict);
+                        decrRefCount(o);
+                        sdsfree(field_dup);
+                        sdsfree(value);
+                        sdsfree(field);
+                        return NULL;
+                    }
+                }
+
+                /* check if the values can be saved to listpack (or should convert to dict encoding) */
+                if (sdslen(field) > server.hash_max_listpack_value ||
+                    sdslen(value) > server.hash_max_listpack_value ||
+                    !lpSafeToAdd(((listpackEx*)o->ptr)->lp, sdslen(field) + sdslen(value) + lpEntrySizeInteger(expire)))
+                {
+                    /* convert to hash */
+                    hashTypeConvert(o, OBJ_ENCODING_HT, hexpires);
+
+                    if (len > DICT_HT_INITIAL_SIZE) { /* TODO: this is NOT the original len, but this is also the case for simple hash, is this a bug? */
+                        if (dictTryExpand(o->ptr, len) != DICT_OK) {
+                            rdbReportCorruptRDB("OOM in dictTryExpand %llu", (unsigned long long)len);
+                            decrRefCount(o);
+                            if (dupSearchDict != NULL) dictRelease(dupSearchDict);
+                            sdsfree(value);
+                            sdsfree(field);
+                            return NULL;
+                        }
+                    }
+
+                    /* don't add the values to the new hash: the next if will catch and the values will be added there */
+                } else {
+                    void *lp = ((listpackEx*)o->ptr)->lp;
+                    lp = lpAppend(lp, (unsigned char *)field, sdslen(field));
+                    lp = lpAppend(lp, (unsigned char *)value, sdslen(value));
+                    lp = lpAppendInteger(lp, expire);
+                    ((listpackEx*)o->ptr)->lp = lp;
```

Comment:
Maybe we can export `listpackExAddNew()` and `listpackExCreate()` and use it in this file.

Reply:
Using `listpackExAddNew` would be, IMHO, not very efficient as it is assumed the hash was already sorted before saving, so each entry read will be added at the end and, so (possibly) the entire listpack has to be traversed for each item. Possibly, a "hint" parameter can be added to it to so that it will be notified the new field should be added as the last entry, and it will only verify this.
WDYT?
(I'll try to use `listpackExCreate` reagrdless)

Reply:
Please correct me if I'm wrong, currently we don't read fields ordered by TTL. So, even without `listpackExAddNew()`, we have to fix the current code here as we don't have items ordered? 

Do we have any other option than calling `listpackExAddNew()`. If field does not have TTL, it will just append the field without traversing it. Otherwise, we have to walk over the list as we have to find the correct spot, no?

Reply:
I guess you're right, order is indeed not guaranteed when reading an object saved as dict, I'll use `listpackExAddNew` isntead.

Reply:
btw a test case for it would be nice. 
e.g. add listpack 100 volatile and 100 non-volatile elements. Verify volatile ones expired after some time. If items are not ordered, active expiry will fail to expire some of the volatile fields.